File: src/libm/numerical_analysis_math/xenly_numerical_analysis_math.c
/*
 * XENLY - high-level and general-purpose programming language
 * created, designed, and developed by Cyril John Magayaga (cjmagayaga957@gmail.com, cyrilmagayaga@proton.me).
 *
 * It is initially written in C programming language.
 *
 * `xenly_numerical_analysis_math.c` is the similar to the `xenly_numerical_analysis_math.jl` in Julia programming language.
 * 
 * It is available for Linux and Windows operating systems.
 *
 */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// GMRES (Generalized Minimal RESidual method)
void xenly_gmres(int n, double A[n][n], double b[n], double x0[n], double tol, int maxiter, double x[n], int *iterations) {
    double Q[n][maxiter + 1], H[maxiter + 1][maxiter];
    double r[n], beta, y[maxiter];
    int i, j;

    // Initialize r, beta, Q
    for (i = 0; i < n; i++) r[i] = b[i] - A[i][0] * x0[0] - A[i][1] * x0[1];
    beta = sqrt(r[0] * r[0] + r[1] * r[1]);
    for (i = 0; i < n; i++) Q[i][0] = r[i] / beta;

    for (j = 0; j < maxiter; j++) {
        double v[n];
        for (i = 0; i < n; i++) v[i] = A[i][0] * Q[0][j] + A[i][1] * Q[1][j];
        for (i = 0; i <= j; i++) {
            H[i][j] = Q[0][i] * v[0] + Q[1][i] * v[1];
            v[0] -= H[i][j] * Q[0][i];
            v[1] -= H[i][j] * Q[1][i];
        }
        H[j+1][j] = sqrt(v[0] * v[0] + v[1] * v[1]);
        if (H[j+1][j] < tol) break;
        for (i = 0; i < n; i++) Q[i][j+1] = v[i] / H[j+1][j];
        for (i = 0; i <= j; i++) y[i] = beta * ((i == j) ? 1 : 0);
        for (i = j; i >= 0; i--) {
            y[i] -= H[i][j] * y[j];
            y[i] /= H[i][i];
        }
        for (i = 0; i < n; i++) x[i] = x0[i] + Q[i][j] * y[j];
        for (i = 0; i < n; i++) r[i] = b[i] - A[i][0] * x[0] - A[i][1] * x[1];
        if (sqrt(r[0] * r[0] + r[1] * r[1]) < tol) {
            *iterations = j;
            return;
        }
    }
    *iterations = maxiter;
}

// Conjugate gradient method
void xenly_conjugate_gradient(int n, double A[n][n], double b[n], double x0[n], double tol, int maxiter, double x[n], int *iterations) {
    double r[n], p[n], Ap[n];
    double alpha, rsold, rsnew;
    int i, iter;

    for (i = 0; i < n; i++) x[i] = x0[i];
    for (i = 0; i < n; i++) r[i] = b[i] - A[i][0] * x0[0] - A[i][1] * x0[1];
    for (i = 0; i < n; i++) p[i] = r[i];
    rsold = r[0] * r[0] + r[1] * r[1];

    for (iter = 0; iter < maxiter; iter++) {
        for (i = 0; i < n; i++) Ap[i] = A[i][0] * p[0] + A[i][1] * p[1];
        alpha = rsold / (p[0] * Ap[0] + p[1] * Ap[1]);
        for (i = 0; i < n; i++) x[i] += alpha * p[i];
        for (i = 0; i < n; i++) r[i] -= alpha * Ap[i];
        rsnew = r[0] * r[0] + r[1] * r[1];
        if (sqrt(rsnew) < tol) {
            *iterations = iter;
            return;
        }
        for (i = 0; i < n; i++) p[i] = r[i] + (rsnew / rsold) * p[i];
        rsold = rsnew;
    }
    *iterations = maxiter;
}

// Gaussian elimination
void xenly_gaussian_elimination(int n, double A[n][n], double b[n], double x[n]) {
    double Ab[n][n+1];
    int i, j, k, maxrow;
    double maxval, factor;

    for (i = 0; i < n; i++) for (j = 0; j < n; j++) Ab[i][j] = A[i][j];
    for (i = 0; i < n; i++) Ab[i][n] = b[i];

    for (i = 0; i < n-1; i++) {
        maxval = fabs(Ab[i][i]);
        maxrow = i;
        for (j = i+1; j < n; j++) {
            if (fabs(Ab[j][i]) > maxval) {
                maxval = fabs(Ab[j][i]);
                maxrow = j;
            }
        }
        
        if (maxrow != i) {
            for (k = 0; k < n+1; k++) {
                double tmp = Ab[i][k];
                Ab[i][k] = Ab[maxrow][k];
                Ab[maxrow][k] = tmp;
            }
        }
        
        for (j = i+1; j < n; j++) {
            factor = Ab[j][i] / Ab[i][i];
            for (k = i; k < n+1; k++) Ab[j][k] -= factor * Ab[i][k];
        }
    }

    for (i = n-1; i >= 0; i--) {
        x[i] = Ab[i][n] / Ab[i][i];
        for (j = i-1; j >= 0; j--) Ab[j][n] -= Ab[j][i] * x[i];
    }
}

// Newton's method
void xenly_newtons_method(void (*f)(double*, double*), void (*df)(double*, double*), double x0[], double tol, int maxiter, double x[], int *iterations) {
    double fx[2], dfx[4], dx[2];
    int i, iter;
    for (i = 0; i < 2; i++) x[i] = x0[i];
    for (iter = 0; iter < maxiter; iter++) {
        f(x, fx);
        if (sqrt(fx[0] * fx[0] + fx[1] * fx[1]) < tol) {
            *iterations = iter;
            return;
        }
        df(x, dfx);
        dx[0] = (fx[0] * dfx[3] - fx[1] * dfx[1]) / (dfx[0] * dfx[3] - dfx[1] * dfx[2]);
        dx[1] = (fx[1] * dfx[0] - fx[0] * dfx[2]) / (dfx[0] * dfx[3] - dfx[1] * dfx[2]);
        for (i = 0; i < 2; i++) x[i] -= dx[i];
    }
    *iterations = maxiter;
}

// Hill Climbing
void xenly_hill_climbing(double (*f)(double*), double x[], double step_size, double tol, int maxiter, double x_out[], double *best_value, int *iterations) {
    double candidate[2], candidate_value;
    int i, iter;

    for (i = 0; i < 2; i++) x_out[i] = x[i];
    *best_value = f(x_out);

    for (iter = 0; iter < maxiter; iter++) {
        for (i = 0; i < 2; i++) candidate[i] = x_out[i] + step_size * (2 * ((double)rand() / RAND_MAX) - 1);
        candidate_value = f(candidate);
        if (candidate_value < *best_value) {
            for (i = 0; i < 2; i++) x_out[i] = candidate[i];
            *best_value = candidate_value;
        }
        
        if (sqrt(candidate_value) < tol) {
            *iterations = iter;
            return;
        }
    }
    *iterations = maxiter;
}

// Bisection method
double xenly_bisection_method(double (*f)(double), double a, double b, double tol, int maxiter, int *iterations) {
    double c;
    int iter;

    for (iter = 0; iter < maxiter; iter++) {
        c = (a + b) / 2;
        if (fabs(f(c)) < tol) {
            *iterations = iter;
            return c;
        }
        
        if (f(a) * f(c) < 0) b = c;
        else a = c;
    }
    *iterations = maxiter;
    return c;
}

// Euler's method
void xenly_eulers_method(double (*f)(double, double), double y0, double t0, double t_end, double h, int *len, double **t_out, double **y_out) {
    int n = (int)((t_end - t0) / h) + 1;
    double *t = (double *)malloc(n * sizeof(double));
    double *y = (double *)malloc(n * sizeof(double));

    int i;
    t[0] = t0;
    y[0] = y0;
    for (i = 1; i < n; i++) {
        t[i] = t[i-1] + h;
        y[i] = y[i-1] + h * f(t[i-1], y[i-1]);
    }
    *len = n;
    *t_out = t;
    *y_out = y;
}

// Example functions
void f(double *x, double *y) {
    y[0] = x[0] * x[0] + x[1] * x[1] - 4;
    y[1] = x[0] * x[0] - x[1] * x[1] - 1;
}

void df(double *x, double (*dy)[2]) {
    dy[0][0] = 2 * x[0];
    dy[0][1] = 2 * x[1];
    dy[1][0] = 2 * x[0];
    dy[1][1] = -2 * x[1];
}

double f_hill_climbing(double x[2]) {
    return (x[0] - 1) * (x[0] - 1) + (x[1] - 1) * (x[1] - 1);
}

double f_bisection(double x) {
    return x * x - 2;
}

double f_euler(double t, double y) {
    return t + y;
}

File: src/libm/numerical_analysis_math/xenly_numerical_analysis_math.jl
#
# XENLY - high-level and general-purpose programming language
# created, designed, and developed by Cyril John Magayaga (cjmagayaga957@gmail.com, cyrilmagayaga@proton.me).
# 
# It is initially written in Julia programming language.
#
# `xenly_numerical_analysis_math.jl` is the similar to the `xenly_numerical_analysis_math.c` in C programming language.
#
module xenly_numerical_analysis_math

export xenly_gmres, xenly_conjugate_gradient, xenly_gaussian_elimination, xenly_newtons_method, xenly_hill_climbing, xenly_bisection_method, xenly_eulers_method

# GMRES (Generalized Minimal RESidual method)
function xenly_gmres(A, b, x0, tol=1e-6, maxiter=100)
    n = length(b)
    Q = Matrix{Float64}(undef, n, maxiter + 1)
    H = zeros(Float64, maxiter + 1, maxiter)
    r = b - A*x0
    beta = norm(r)
    Q[:, 1] = r / beta

    for j = 1:maxiter
        v = A*Q[:, j]
        for i = 1:j
            H[i, j] = dot(Q[:, i], v)
            v -= H[i, j] * Q[:, i]
        end
        H[j+1, j] = norm(v)
        if H[j+1, j] < tol
            break
        end
        Q[:, j+1] = v / H[j+1, j]
        y = H[1:j, 1:j] \ (beta * I(j))
        x = x0 + Q[:, 1:j] * y
        r = b - A*x
        if norm(r) < tol
            return x, j
        end
    end

    return x, maxiter
end

# Conjugate gradient method
function xenly_conjugate_gradient(A, b, x0, tol=1e-6, maxiter=100)
    n = length(b)
    x = copy(x0)
    r = b - A * x
    p = copy(r)
    rsold = dot(r, r)

    for iter = 1:maxiter
        Ap = A * p
        alpha = rsold / dot(p, Ap)
        x += alpha * p
        r -= alpha * Ap
        rsnew = dot(r, r)
        if sqrt(rsnew) < tol
            return x, iter
        end
        p = r + (rsnew / rsold) * p
        rsold = rsnew
    end

    return x, maxiter
end

# Gaussian elimination
function xenly_gaussian_elimination(A, b)
    n = size(A, 1)
    Ab = hcat(A, b)

    for i = 1:n-1
        # Pivot selection (partial pivoting)
        maxval, maxrow = findmax(abs.(Ab[i:end, i]))
        maxrow += i - 1

        if maxval == 0
            error("Matrix is singular or nearly singular.")
        end

        if maxrow != i
            Ab[[i, maxrow], :] = Ab[[maxrow, i], :]
        end

        for j = i+1:n
            factor = Ab[j, i] / Ab[i, i]
            Ab[j, i:end] -= factor * Ab[i, i:end]
        end
    end

    x = zeros(n)
    x[n] = Ab[n, n+1] / Ab[n, n]

    for i = n-1:-1:1
        x[i] = (Ab[i, n+1] - dot(Ab[i, i+1:n], x[i+1:n])) / Ab[i, i]
    end

    return x
end

# Newton's method
function xenly_newtons_method(f, df, x0, tol=1e-6, maxiter=100)
    x = copy(x0)
    iter = 0
    while norm(f(x)) > tol && iter < maxiter
        x -= df(x) \ f(x)
        iter += 1
    end
    return x, iter
end

# Hill Climbing
function xenly_hill_climbing(f, x0; step_size=0.01, tol=1e-6, maxiter=1000)
    x = copy(x0)
    best_value = f(x)
    iter = 0
    while iter < maxiter
        candidate = x + step_size * (2 * rand(length(x)) .- 1)
        candidate_value = f(candidate)
        if candidate_value < best_value
            x = candidate
            best_value = candidate_value
        end
        iter += 1
        if norm(candidate - x) < tol
            break
        end
    end
    return x, best_value, iter
end

# Bisection method
function xenly_bisection_method(f, a, b; tol=1e-6, maxiter=1000)
    if f(a) * f(b) > 0
        error("f(a) and f(b) must have opposite signs")
    end

    fa, fb = f(a), f(b)
    iter = 0
    while (b - a) / 2 > tol && iter < maxiter
        c = (a + b) / 2
        fc = f(c)
        if fc == 0.0
            return c, iter
        elseif fa * fc < 0
            b = c
            fb = fc
        else
            a = c
            fa = fc
        end
        iter += 1
    end

    return (a + b) / 2, iter
end

# Euler's method
function xenly_eulers_method(f, y0, t0, t_end, h)
    t = t0:h:t_end
    y = zeros(length(t))
    y[1] = y0
    for i = 2:length(t)
        y[i] = y[i-1] + h * f(t[i-1], y[i-1])
    end
    return t, y
end

end # module